onnxruntime/core/providers/xnnpack/nn/unpool.h

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+/*
+xnnpack_unpool can only work with xnnpack_maxpool, its index is totally different from onnxruntime's.
```

Comment:
Can you clarify what's supported given we use an xnnpack kernel for MaxPool but not here. If the indices are totally different how can we use an xnnpack kernel for MaxPool? Wouldn't any nodes that consume those indices (other than a MaxUnpool running on xnnpack) be broken?

Reply:
Sorry, this comment should be removed with the relevant code of xnn_max_unpool_create/setup/run.

`Maxunpool` is a little bit special, it consumes NHWC input data and NCHW indice (2 inputs for this Op), generates NHWC output.

A big concern is that we don't have a definition of MaxPool in NHWC layout, how to arrange the second output.

This OP works for
`NCHW MAXPOOL`--->`Transpose the first output to NHWC`----------->`NHWC MaxUnpool`.
............................. \\--->`keep the second output Indice unchanged`---/
If it's unreasonable, I would rather undo this operator. How do you think?

Reply:
If you are converting a MaxUnpool from the original NCHW layout to NHWC don't you have to insert a Transpose node on the indicies input as well? 

So whether the MaxPool was done in NCHW or NHWC shouldn't matter.

e.g. 
X(nchw) -> MaxPool -> Y(nchw) and Indicies(nchw)  -> ... -> Y(nchw), Indicies(nchw) -> MaxUnpool -> Z(nchw)

If I want to convert the MaxUnpool to NHWC the layout transformer should be transposing both Y and Indicies, as well as the output

Y(nchw)           -> Transpose -> Y(nhwc)           -> MaxUnpool(nhwc) -> Z(nhwc) -> Transpose -> Z(nchw)
Indicies(nchw) -> Transpose -> Indicies(nhwc) ->

On the boundary of the Transpose nodes the input is NCHW and the output is NCHW. 

If the layout optimizer is not currently transposing the Indicies value it should be fixed.

If there's a MaxPool prior the same approach applies

original
X(nchw) -> MaxPool(nchw) -> Y(nchw), Indicies(nchw)
layout converted
X(nchw) -> Transpose -> X(nhwc) -> MaxPool(nhwc) -> Y(nhwc)           -> Transpose -> Y(nchw)
                                                                                     -> Indicies(nhwc) -> Transpose -> Indicies(nchw)  

Once we plug in the NHWC versions of the pooling ops we can optimize the Transpose nodes. If we _only_ have MaxUnpool(nhwc) we can't change anything. If we have both MaxPool(nhwc) and MaxUnpool(nhwc) we can most likely get rid of the Transpose nodes between them as the data can stay in NHWC layout (unless in the middle is a layout sensitive operator that we do not have an NHWC kernel for, at which point it has to go back to NCHW). 

Either way, if we do layout changes correctly an operator should never be getting layout sensitive inputs in a mix of layouts.

Reply:
Actually given the indicies are flattened value it would require something other than a Transpose to convert between layouts. Probably do-able but question is how much if any performance benefit we'd get. Might depend on how often we're pooling/unpooling and what's in between.

Reply:
Alternatively if you have a MaxPool node you _could_ convert to NHWC and the Indicies output is directly connected to a MaxUnpool you can also convert to NHWC, and there are no usages of the Indicies value in between (can check the consumer count for the value), it might be safe to convert both as the indicies format really becomes internal to those two nodes.

Reply:
A relative `Indices`, which used in XNNPACK, is more general. Might be possible support by ONNX spec someday. 0.0

I would revert this Op since it's not a very common one. won't add it back until it's required in a specific model.